compiler/rustc_passes/src/dead.rs

```diff
@@ -633,10 +680,17 @@ impl<'tcx> DeadVisitor<'tcx> {
     fn should_warn_about_field(&mut self, field: &hir::FieldDef<'_>) -> bool {
         let def_id = self.tcx.hir().local_def_id(field.hir_id);
         let field_type = self.tcx.type_of(def_id);
-        !field.is_positional()
-            && !self.symbol_is_live(def_id)
+        !self.symbol_is_live(def_id)
             && !field_type.is_phantom_data()
-            && !has_allow_dead_code_or_lang_attr(self.tcx, field.hir_id)
+            && !(if field.is_positional() {
+                has_allow_dead_code_or_lang_attr_helper(
+                    self.tcx,
+                    field.hir_id,
+                    lint::builtin::UNUSED_TUPLE_STRUCT_FIELDS,
+                )
+            } else {
+                has_allow_dead_code_or_lang_attr(self.tcx, field.hir_id)
+            })
```

Comment:
I don't love if expressions that resolve to bool, but I don't have a good reason to ask you to change it ^_^'